k-distribution/include/kframework/builtin/ffi.md

```diff
@@ -0,0 +1,177 @@
+corrupted term representations that lead to undefined behavior at runtime.
+
+```k
+require "domains.k"
+
+module FFI-SYNTAX
+  imports LIST
+```
+
+The FFIType sort is used to declare the native C ABI types of operands passed
+to the `#ffiCall` function. These types roughly correspond to the types 
+declared in `ffi.h` by libffi.
+
+```k
+  syntax FFIType ::= "#void" [klabel(#void), symbol]
+                  | "#uint8" [klabel(#uint8), symbol]
+                  | "#sint8" [klabel(#sint8), symbol]
+                  | "#uint16" [klabel(#uint16), symbol]
+                  | "#sint16" [klabel(#sint16), symbol]
+                  | "#uint32" [klabel(#uint32), symbol]
+                  | "#sint32" [klabel(#sint32), symbol]
+                  | "#uint64" [klabel(#uint64), symbol]
+                  | "#sint64" [klabel(#sint64), symbol]
+                  | "#float" [klabel(#float), symbol]
+                  | "#double" [klabel(#double), symbol]
+                  | "#uchar" [klabel(#uchar), symbol]
+                  | "#schar" [klabel(#schar), symbol]
+                  | "#ushort" [klabel(#ushort), symbol]
+                  | "#sshort" [klabel(#sshort), symbol]
+                  | "#uint" [klabel(#uint), symbol]
+                  | "#sint" [klabel(#sint), symbol]
+                  | "#ulong" [klabel(#ulong), symbol]
+                  | "#slong" [klabel(#slong), symbol]
+                  | "#longdouble" [klabel(#longdouble), symbol]
+                  | "#struct" "(" List ")" [klabel(#struct), symbol]
+endmodule
+
+module FFI
+  imports FFI-SYNTAX
+  imports BYTES
+  imports STRING
+
+```
+
+FFI Calls
+---------
+
+The `#ffiCall` functions are designed to call a native C ABI function and 
+return a native result. They come in three variants:
+
+### Non-variadic
+
+In the first variant, `#ffiCall(Address, Args, ArgTypes, ReturnType)` takes
+an integer address of a function (which can be obtained from
+`#functionAddress`), a `List` of `Bytes` containing the arguments of the
+function, a `List` of `FFIType`s containing the types of the parameters of the
+function, and an `FFIType` containing the return type of the function, and 
+returns the return value of the function as a `Bytes`.
+
+```k
+  syntax Bytes ::= "#ffiCall" "(" Int "," List "," List "," FFIType ")" [function, hook(FFI.call)]
+```
+
+### Variadic
+
+In the second variant,
+`#ffiCall(Address, Args, FixedTypes, VariadicTypes, ReturnType` takes an
+integer address of a function, a `List` of `Bytes` containing the arguments
+of the call, a `List` of `FFIType`s containing the types of the fixed
+parameters of the function, a `List` of `FFIType`s containing the types of the
+variadic parameters of the function, and an `FFIType` containing the return
+type of the function, and returns the return value of the function as a
+`Bytes`.
+
+```k
+                 | "#ffiCall" "(" Int "," List "," List "," List "," FFIType ")" [function, hook(FFI.call_variadic)]
+```
```

Comment:
```suggestion
```k
  syntax Bytes ::= "#ffiCall" "(" Int "," List "," List "," List "," FFIType ")" [function, hook(FFI.call_variadic)]
```
```

Reply:
Otherwise it's confusing and you're not able to copy-paste this code block around to other locatiions.